File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:

    vector<vector<int>> res;
    vector<int> path; //subnets right now

    void backtrack(int start, vector<int> &nums) {
        // Save the current subset
        res.push_back(path);

        for (int i = start; i < nums.size(); ++i) {
            path.push_back(nums[i]);
            backtrack(i + 1, nums);
            path.pop_back();
        }
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        backtrack(0, nums);
        return res;
    }
};

//TimeCP: O(n*2^n)
//SpaceCP: O(n*2^n) include output

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        dem, tich =  0, 1

        for i in range(len(nums)):
            if nums[i] == 0:
                dem += 1
            else: tich *= nums[i]

        res = []

        if dem == 0:
            for i in range(len(nums)):
                res.append(tich // nums[i])
            return res
                
        elif dem == 1:
            for i in range(len(nums)):
                if nums[i] == 0:
                    res.append(tich)
                else: res.append(0)
            return res
        
        else: 
            return [0] * len(nums)
